### PR TITLE
Port BannedApiAnalyzer

### PR DESCRIPTION
**Pull request by sharwell**
Port BannedApiAnalyzer from dotnet/roslyn-analyzers

Only minimal cosmetic changes were made to account for the new test framework, some StyleCop Analyzers rule violations, and documentation requirements.

### CODE REVIEW

File: BannedApiAnalyzer/BannedApiAnalyzer.CSharp/ApiDesignRules/CSharpSymbolIsBannedAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Tunnel Vision Laboratories, LLC. All Rights Reserved.
```
Comment (genlu) [inline BannedApiAnalyzer/BannedApiAnalyzer.CSharp/ApiDesignRules/CSharpSymbolIsBannedAnalyzer.cs:1]: Just want to call out the different of copyright and license info. I'm not a lawyer, would this be a problem?

```
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```

Comment (sharwell) [reply BannedApiAnalyzer/BannedApiAnalyzer.CSharp/ApiDesignRules/CSharpSymbolIsBannedAnalyzer.cs:1]: The attribution for the original content is here:

https://github.com/DotNetAnalyzers/BannedApiAnalyzer/blob/a6a45026fb5ea6121deac6e51b329cbe2c3181ac/THIRD-PARTY-NOTICES.txt#L10-L24

File: BannedApiAnalyzer/BannedApiAnalyzer/ApiDesignRules/ApiDesignResources.resx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    <value>2.0</value>
+  </resheader>
+  <resheader name="reader">
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="DuplicateBannedSymbolDescription" xml:space="preserve">
+    <value>The list of banned symbols contains a duplicate.</value>
```
Comment (genlu) [inline BannedApiAnalyzer/BannedApiAnalyzer/ApiDesignRules/ApiDesignResources.resx:121]: All the text are localized in roslyn-analyzer repo. Are we just gonna drop it?

Comment (sharwell) [reply BannedApiAnalyzer/BannedApiAnalyzer/ApiDesignRules/ApiDesignResources.resx:121]: > All the text are localized in roslyn-analyzer repo. Are we just gonna drop it?

No, I can bring over the localized strings. I didn't realize localization was already completed for this analyzer.

**Comment by genlu** [comment]
We need to port the VersionCheckAnalyzer as well, since this is backed by IOperation.

**Comment by genlu** [comment]
There's no CodeFix support for this analyzer, should remove codefix related stuff from this repo. If there's any future work planned, then at least remove them from nuget package and vsix for now.
This could be done after this is merged.

**Review by genlu — APPROVED** [review]
Thanks @sharwell!

File: BannedApiAnalyzer/BannedApiAnalyzer/ApiDesignRules/ApiDesignResources.resx (hunk 2 above)
Comment (sharwell) [reply BannedApiAnalyzer/BannedApiAnalyzer/ApiDesignRules/ApiDesignResources.resx:121]: ➡️ Fixed now

**Comment by sharwell** [comment]
> We need to port the VersionCheckAnalyzer as well, since this is backed by IOperation.

It references Roslyn 2.6.0, at which point IOperation became a stable public API. I don't believe there is any other need for the VersionCheckAnalyzer here.

**Comment by sharwell** [comment]
> There's no CodeFix support for this analyzer, should remove codefix related stuff from this repo. If there's any future work planned, then at least remove them from nuget package and vsix for now.

This is something we can revisit in the future, but the current packaging won't be problematic either way.

**Comment by genlu** [comment]
@sharwell this analyzer won't work with older version of VS, right (either as vsix or NuGet)? Isn't the purpose of VersionCheckAnalyzer to generate a sensible warning to the user in that scenario?

**Pull request merged**